📚 [Guide] Create a backward compatible Turbomodule (#3023)
facebook/react-native-website: docs/new-architecture-backward-turbomodule.md

---
id: new-architecture-backward-compatible-turbomodule
title: Create a Backward Compatible Turbomodule
---

import NewArchitectureWarning from './\_markdown-new-architecture-warning.mdx';
import M1Cocoapods from './\_markdown-m1-cocoapods.mdx';

<NewArchitectureWarning/>

> This guide describes how to create a Turbomodule, by keeping it backward compatibile with the old Architecture. To achieve this result, we first need to create a Native Module and then to build a Turbomodule on top of it.

Author: Section added! 👍

## What Is a Turbomodule, In Short.

Turbomodules are the evolution of Native Modules. The main benefit of using a Turbomodule is that the communication between the native layer and the JavaScript layer happens using shared memory instead of the bridge. This allows for better performances and synchronous calls between the native code and JavaScript, while the bidge forced us to work with asynchronous communication where all the messages were serialized using a JSON format.

The communication is handled by a C++ layer and this code is generated automatically from the JS code. Native developers don't have to write C++, if not for some very small bits. The C++ code is shared across all the supported platforms (iOS, Android, ...): we don't have different implementation for each of them. This improve maintainability and uniformity across platform.

> <!-- To learn more about Turbomodule, check this [deep dive](). -->

Author: Agreed. In the final version we will have different sections to refer to.

## Create the Native Module

This section and the following are a practical guides to create a module that is backward compatible. We are not diving deep into what can be done with Native Modules. For more informations on that topic, have a look [here](native-modules-intro).

> If you are interested in creating a Turbomodule, you can skip this part and directly go [here](#create-the-turbomodule). Consider that most of the code of a Native Module is needed for a Turbomodule.

Author: I agree. I moved the sentence to the previous paragraph, adding some more context.

### Creating the Structure

We have two alternatives to create the basic structure for our module.

1. Using the React Native CLI
2. Manually.

#### React Native CLI

To create a turbomodule with the React Native CLI, run the following command:

> ```sh title=""

Reviewer: ```suggestion
```sh
```

Just a small fix suggestion.

npx create-react-native-library <name-of-the-library>
```

Follow the questions the CLI asks and, once the process is done, it will create a folder structure completed with an example app, typescript configuration and much more. The sample structure is shown in the following image:

<figure>
  <img src="docs/assets/Turbomodules/NpxLibFolderStructure.png" />
  <figcaption>The folder structure generated by the React Native CLI</figcaption>
</figure>

This is a good starting point, but remember to double check what will be published if you keep this setup. Moreover, the `TypeScript` support for the new architecture is still in beta, so you won't be able to use it for the Turbomodule sections.

#### Manually

The second approach is to create a similar structure manually. What we need to start is a folder named `example-library` and a `package.json` file with the following content:

```json title="package.json"
{
  "name": "example-library",
  "version": "0.0.1",

>   "description": "Showcase Turbomodule with Backcompatibility ",

Reviewer 2: ```suggestion
  "description": "Showcase Turbomodule with Backward compatibility",
```

  "react-native": "src/index",
  "source": "src/index",
  "files": [
    "src",
    "android",
    "ios",
    "example-library.podspec",
    "!android/build",
    "!ios/build",
    "!**/__tests__",
    "!**/__fixtures__",
    "!**/__mocks__"
  ],
  "keywords": ["react-native", "ios", "android"],
  "repository": "https://github.com/<your_github_handle>/example-library",
  "author": "<Your Name> <your_email@your_provide.com> (https://github.com/<your_github_handle>)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/<your_github_handle>/example-library/issues"
  },
  "homepage": "https://github.com/<your_github_handle>/example-library#readme",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/"
  },
  "devDependencies": {},
  "peerDependencies": {
    "react": "*",
    "react-native": "*"
  }
}
```

For sake of simplicity, we will go on with this manual setup, adding files while we work.
Remember to update the `<Your Name>`, `<your_email@your_provide.com>`, and all the `<your_github_handle>` tokens, if you want to follow this approach with your information.

### Create a Testing Environment

> To develop a Native Module we would like to test the JS side and to develop safely the native parts. A library usually does not come with a testing aplication out-of-the-box. So, we are going to need one.

Reviewer 2: > we would like to test the JS side and to develop safely the native parts

I'm not sure I follow here. Why you want to test ONLY the JS side. And what does "develop safely" means in this context?

> To develop a Native Module we would like to test the JS side and to develop safely the native parts. A library usually does not come with a testing aplication out-of-the-box. So, we are going to need one.

Reviewer 2: ```suggestion
To develop a Native Module we would like to test the JS side and to develop safely the native parts. A library usually does not come with a sample application out-of-the-box. So, we are going to create one.
```

:::note
If you used the React Native CLI to create the module, you have an application called `example` within you `example-library` folder.
:::

To create a proper test environment, let's navigate to the parent folder of our `example-library` and run the following commands:

```sh
npx react-native init OldArchitectureApp
cd OldArchitectureApp
yarn install
```

Now, open a new terminal and start Metro by running:

```sh

> npx react-native start

Reviewer 2: Please note that we're mixing `npx` and `yarn` here. That's not idea. I would stick to only yarn at this point.

Author: I'm not sure we can do that. 
`yarn` does not support to init a RN app as easily as `npx`. Also, following the [Setting up your environment](https://reactnative.dev/docs/environment-setup), it looks like that `yarn install` is not really needed. I would stick to `npx`, coherently with the other sections of the website! 👍 

Still... we will use yarn to install packages, because the dependencies are handled in that way.

```

Go back to previous terminal and run:

> ```sh
> npx react-native run-ios
> ```

Reviewer 2: ```suggestion
```sh
npx react-native run-ios
# Or for Android
npx react-native run-android
```
```

Also some of those basic commands can be omitted and we can assume a basic knowledge of the RN CLI

When all the processes are done, you should see the default React Native application, running on a simulator.
We will update the code of this app to test our Native Module.

### Define the JS Shared Code

The next step is to define how the React Native app will use our module. Native modules are registered using the `NativeModule` object provided by the React Native framework.

For this example, we are going to create a manager that asks the GPS authorization to the user.

As first step, let's create a folder named `src` and an `index.js` file into it. Add the following code into the JS file.

```js title="index.js"
// @flow
// Import the required objects
import { NativeModules, Platform } from 'react-native';

// Declare an error message, triggered when the Native Module is used but not properly linked.
const LINKING_ERROR =
  'The package "AuthorizationManager" doesn\'t seem to be linked. Make sure: \n\n' +
  Platform.select({
    ios: "- You have run 'pod install'\n",
    default: ''
  }) +
  '- You rebuilt the app after installing the package\n' +
  '- You are not using Expo managed workflow\n';

> // Create a Proxy object that raises the error if the module is not linked.

Reviewer 2: Is the proxy object needed?

Author: Not really. It is something create by template that I find helpful to get some clear error message.
I agree, it makes things more complex. I'll remove it.

const manager = NativeModules.AuthorizationManager
  ? NativeModules.AuthorizationManager
  : new Proxy(
      {},
      {
        get() {
          throw new Error(LINKING_ERROR);
        }
      }
    );

// Export the required types.
export type AuthorizationStatus =
  | 'NotDetermined'
  | 'Restricted'
  | 'Denied'
  | 'Authorized'
  | 'AuthorizedAlways'
  | 'AuthorizedWhenInUse';

export default manager;
```

:::info
This file uses [`Flow`](https://flow.org/) to work with types.
Types are required for the Turbomodules.
:::

### Implement the iOS Code

Let's create an `ios` folder and then open Xcode. Then, follow these steps:

1. In the Apple's IDE, create a new project.
2. Select the `iOS` tab and the `Static Library` template.
3. Choose a name and a `bundleId`. For this project, let's use `Objective-C` as language.
4. Select the `example-library/ios` folder as destination. Remember to uncheck the `create git repository` option.

Xcode will create a small project with two files: the `ExampleLibrary.h` and the `ExampleLibrary.m`. To stay faithful to Apple convention, let's rename them in `ELExampleLibrary.h` and `ELExampleLibrary.m`.

:::important

> The `ExampleLibrary.xcodeproj` must be in the `ios` folder. If you put it into a nested folder, module autodiscovery won't be able to find it.

Reviewer 2: autodiscovery -> autolinking

            return @"AuthorizedAlways";
        case kCLAuthorizationStatusAuthorizedWhenInUse:
            return @"AuthorizedWhenInUse";
    }
}

RCT_REMAP_METHOD(getCurrentStatus, getCurrentStatusWithResolver
                 : (RCTPromiseResolveBlock)resolve withRejecter
                 : (RCTPromiseRejectBlock)reject)
{
    CLAuthorizationStatus authStatus = CLLocationManager.authorizationStatus;
    NSString * stringStatus = [self stringStatusFrom:authStatus];
    resolve(stringStatus);
}

RCT_EXPORT_METHOD(askAuthorization)
{
    [manager requestAlwaysAuthorization];
}
```

This code does:

1. Import the proper headers.
1. Declare a private variable to keep a reference to the manager.
1. Initialize the module with a new instance of the manager.
1. Export the module with a custom name.
1. Add a utility method to convert the status in one of the strings we plan to use in JS.
1. Export a method to get the current status.
1. Export a method to request the localization permissions.

For further information about what we can do in iOS with a Native Module, have a look [here](https://reactnative.dev/docs/next/native-modules-ios).

### Implement the Android Code

**TBD**

> For further information about what we can do in iOS with a Native Module, have a look [here](https://reactnative.dev/docs/next/native-modules-android).

Reviewer 2: ```suggestion
For further information about what we can do in Android with a Native Module, have a look [here](https://reactnative.dev/docs/next/native-modules-android).
```

 * @flow strict-local
 */

import React from 'react';
import { useEffect, useState } from 'react';
import type { Node } from 'react';
import {
  SafeAreaView,
  StatusBar,
  Text,
  View,
  Button
} from 'react-native';

import type { AuthorizationStatus } from 'example-library/src/index';
import AuthorizationManager from 'example-library/src/index';

const App: () => Node = () => {
  const [authorization, setAuthorization] =
    useState<AuthorizationStatus>('NotDetermined');

  useEffect(() => {
    async function setInitialState() {
      const newState =
        await AuthorizationManager.getCurrentStatus();
      setAuthorization(newState);
    }
    setInitialState();
  }, []);

  return (
    <SafeAreaView>
      <StatusBar barStyle="dark-content" />

      <Button
        title="Request Authorization"
        onPress={() => {
          AuthorizationManager.askAuthorization();
        }}
      />

      <Text style={{ marginLeft: 20 }}>
        Current authorization is: {authorization}
      </Text>
    </SafeAreaView>
  );
};

export default App;
```

We can now go back to the terminal and start Metro:

```sh
npx react-native start
```

and then run the app:

```sh

> npx react-native run-ios

Reviewer 2: either Android or iOS

```

After the app launches, we should see the following state in the simulator:

<figure>
  <img src="docs/assets/Turbomodules/OldArchApp.png" />
  <figcaption>The simulator running the app</figcaption>
</figure>

But, if we tap on the button, nothing happens. That's because the `Core Location` system requires some keys in the plist. Let's do the following:

1. Open the `ios/OldArchitectureApp.xcworkspace`.
1. Select the project in the project navigator.
1. Select the info tab.
1. Scroll down and add the following two keys:
   - `NSLocationWhenInUseUsageDescription`
   - `NSLocationAlwaysAndWhenInUseUsageDescription`

> 1. Add any description you want for each of these.

Reviewer 2: This is really specific to `Core Location`. As this page is already long, shouldn't we try to simplify the TM to do something easier (say write on the log)?

I know it won't be as realistic as this example, but it would be easier to consume as a guide.

Author: Yeah. I thought about this. I wanted to find an example that makes sense for both iOS and Android. 
Unfortunately, `RCTLog` does not work with Hermes and Turbomodules and it limits a little bit the possibilities to showcase what we can do with a TM.
What about a calculator, instead? So we can show also return types an parameters? 🤔

</figure>

Congratulation! The Native Module is now correctly connected to your app.
If you now relaunch the app, you should see the status getting updated.

:::warning
This is an example. Don't use this module for your location-based features.
:::

## Create the Turbomodule

Let's now see how to create a Turbomodule. A Turbomodule is defined by a set of JS spec that needs to be typed. For that reason, we are going to use `Flow` dialect. `TypeScript` support is still in beta but we hope to release it soon.

Once we have the JS spec, we can proceed generating the C++ code that bridges the native code with the JS counterpart.

For the final step, we will implement the missing bits in the native code and then we will connect it to JS.

But first, let's create a testing environment suitable for the new architecture.

### Create the New Architecture Testing Environment

The steps to create a New Architecture testing environment are pretty much the same as [here](#create-a-testing-environment).

There are only two differences. The first one is how we create the app. We will need to run:

```sh
npx react-native init NewArchitectureApp --version next
```

> The `--version next` bit will let us retrieve a version of React Native that is compatible with the new architecture. We can then proceed with:

Reviewer 2: That's risky imho. This guide will then change behavior based on when you execute it. 

Ideally we should mention that version should be at least 0.68 or so.

Author: I tried to fix this. I kept `--version next` in the command because I couldn't find how to explicitly use the `0.68` release candidate in it, but I added some explanation in the line below. Let me know if it works! ;)

```sh
yarn install
```

The second difference arrives here, with the pod installation. Instead of running the usual `pod install`, we should run:

```sh
cd ios && RCT_NEW_ARCH_ENABLED=1 pod install && cd ..
```

The `RCT_NEW_ARCH_ENABLED` flag will tell to cocoapods to install all the dependencies required by the New Architecture. As final step, let's start Metro in a different terminal:

```sh
npx react-native start
```

And let's run the app, to make sure that everything works:

```sh
npx react-native run-ios
```

### Adding the JS Spec

> Let's add the Turbomodule JS specs. As first thing, we will extract the `AuthorizationStatus` in its own file, to share it between the two modules.

Reviewer 2: Couldn't we just let the user create two separate files beforehand?

Author: Done! :D

Let's create a new `AuthorizationStatus.js` file in the `src` folder and move this code from the `index.js` file.

> ```js title="AuthorizationStatus.js"

Reviewer 2: ```suggestion
```ts title="AuthorizationStatus.js"
```

We use `ts` for Flow files for better syntax highlightining

// @flow
export type AuthorizationStatus =
  | 'Unknown'
  | 'Denied'
  | 'Ephemeral'
  | 'Authorized'
  | 'Provisional';
```

> Let's now create a new js file called `NativeAuthorizationManager.js`. The `Native` prefix is **mandatory** for the codegen to work. The code generation process will look for all the modules with the `Native` prefix and it will extract the specs from there. Let's populate the file with the following code:

Reviewer 2: > The `Native` prefix is **mandatory** for the codegen to work.

Maybe this inside a warning box?

> ```js title="NativeAuthorizationManager.js"

Reviewer 2: ```suggestion
```ts title="NativeAuthorizationManager.js"
```

```json title="example-library/package.json"
    "codegenConfig": {
        "libraries": [
        {
            "name": "ELAuthorizationManagerSpec",
            "type": "modules",
            "jsSrcsDir": "src"
        }
    ]
  }
```

The `name` property shows how the spec file will be named in the native code. The `type` property states that we want to generate the code for a Turbomodule. Finally, the `jsSrcsDir` indicates where the specs are located.

#### Add the Turbomodule to an app
Before adding the Turbomodule to the app, we need to update the `example-library.podspec`. Let's go to the `example-library` folder and open the `example-library.podspec`.

Before the `Pod::Spec.new`, add the following variables:

```ruby title="example-library.podspec"
folly_version = '2021.06.28.00-v2'
folly_compiler_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -Wno-comma -Wno-shorten-64-to-32'
```

Before the `end` keyword, add the following snippet:

```ruby title="example-library.podspec"
  if ENV['RCT_NEW_ARCH_ENABLED'] == '1' then
    s.compiler_flags = folly_compiler_flags + " -DRCT_NEW_ARCH_ENABLED=1"

>     s.pod_target_xcconfig    = {

Reviewer 2: ```suggestion
    s.pod_target_xcconfig = {
```

      "HEADER_SEARCH_PATHS" => "\"$(PODS_ROOT)/boost\""
    }

    s.dependency "React-Codegen"
    s.dependency "RCT-Folly", folly_version
    s.dependency "RCTRequired"
    s.dependency "RCTTypeSafety"
    s.dependency "ReactCommon/turbomodule/core"
  end
```

> Practically, this snippet adds some configuration and dependencies required by the New Architecture in order to work properly. These are guarded by the `if ENV['RCT_NEW_ARCH_ENABLED'] == '1' then` control statement: when we install the pods for the new architecture, we use the `RCT_NEW_ARCH_ENABLED` flag. By checking that this flag isenabled in the `podspec`, we install these additional dependencies only for that architecture.

Reviewer 2: ```suggestion
Practically, this snippet adds some configuration and dependencies required by the New Architecture in order to work properly. These are guarded by the `if ENV['RCT_NEW_ARCH_ENABLED'] == '1' then` control statement: when we install the pods for the new architecture, we use the `RCT_NEW_ARCH_ENABLED` flag. By checking that this flag is enabled in the `podspec`, we install these additional dependencies only for that architecture.
```

Now, we can navigate to the `NewArchitectureApp` and run the usual `yarn add ../example-library` to add the module to the app.

#### Install the pods [iOS]

Finally, we can navigate to the `NewArchitectureApp` folder and run the command:

```sh
cd ios && RCT_NEW_ARCH_ENABLED=1 pod install && cd ..
```

If everything goes well, you should be able to see this lines in the terminal:

```sh
[Codegen] >>>>> Processing ELAuthorizationManagerSpec
[Codegen] Generated schema: /var/folders/b7/5gvyd0914t15w42kwy1k_l600000gn/T/ELAuthorizationManagerSpec1DIhKQ/schema.json
[Codegen] Generated artifacts: /Users/cipolleschi/Exmples/NewArch/ios/build/generated/ios/ELAuthorizationManagerSpec
```

In the `ios` folder of the app, a new folder called `build` should appear. The folder contains all the generated code. Among them, you should find a folder called `ELAuthorizationManagerSpec` which contains two files: `ELAuthorizationManagerSpec.h` and `ELAuthorizationManagerSpec-generated.mm`.

> The former is the interface for our type. The latter contains the implementation and the code that connects the native wolrd with the JS one.

Reviewer 2: ```suggestion
The former is the interface for our type. The latter contains the implementation and the code that connects the native world with the JS one.
```

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params
{
  return std::make_shared<facebook::react::NativeAuthorizationManagerSpecJSI>(params);
}
#endif
```

This code implements a new function of the Turbomodule architecture that lets React Native create a concrete implementation of the JS specs. Once again, we are guarding that code with the `#ifdef RCT_NEW_ARCH_ENABLED` conditional compilation pragma to avoid to compile it in the old architecture.

Let's build the app to make sure that everything works and that there are no building issues.

:::warning
Changes in Xcode won't be reflected in the right folder. The pod is installed from the `node_modules` folder and not from the proper `example-library` folder.
Once everything builds, remember to copy the code we modified in Xcode to the `example-library/ELExampleLibrary.mm` file in the `example-library` folder.
:::

### Update the Android Code

**TBD**

### Test the Turbomodule

The last step of this guide consist in testing that our `NewArchitectureApp` works with our Turbomodule.

We already run some of the steps to connect the module with the app.

> Open the `NewArchitectureApp/App.js` and replace the code with the same one we used in the `OldArchitectureApp`.

Reviewer 2: You can use a ````diff` block to explain which lines needs to be changed.

        await AuthorizationManager.getCurrentStatus();
      setAuthorization(newState);
    }
    setInitialState();
  }, []);

  return (
    <SafeAreaView style={backgroundStyle}>
      <StatusBar
        barStyle={isDarkMode ? 'light-content' : 'dark-content'}
      />
      <Button
        title="Request Authorization"
        onPress={() => {
          AuthorizationManager.askAuthorization();
        }}
      />
      <Text>Current authorization is: {authorization}</Text>
    </SafeAreaView>
  );
};

export default App;
```

Now, the only things we have to change are the `import` statements. Replace the `import type { AuthorizationStatus }` and the `import AuthorizationManager` lines with the following:

```js title="App.js imports"
import type { AuthorizationStatus } from 'example-library/src/AuthorizationStatus';
import AuthorizationManager from 'example-library/src/NativeAuthorizationManager';
```

Before running the app, remember to add the required `Info.plist` keys:

- `NSLocationWhenInUseUsageDescription`
- `NSLocationAlwaysAndWhenInUseUsageDescription`

Now, from the terminal, run the `npx react-native run-ios` command to see the app working.

## Summary

> To summarise, the steps to create a Turbomodule which is bacward compatible with the old architecture, are:

Reviewer 2: ```suggestion
To summarise, the steps to create a Turbomodule which is backward compatible with the old architecture, are:
```

---
id: new-architecture-backward-compatible-turbomodule
title: Create a Backward Compatible Turbomodule
---

<NewArchitectureWarning/>

This guide describes how to create a Turbomodule, by keeping it backward compatibile with the old Architecture. To achieve this result, we first need to create a Native Module and then to build a Turbomodule on top of it.

> The guide is organized as it follows:
> 
> 1. [What is a Turbomodule, In Short](#what-is-a-turbomodule-in-short)
> 1. [Create the Native Module](#create-the-native-module)
>    1. [Creating the Structure](#creating-the-structure)
>    1. [Create a Testing Environment](#create-a-testing-environment)
>    1. [Define the JS Shared Code](#define-the-js-shared-code)
>    1. [Implement the iOS Code](#implement-the-ios-code)
>    1. [Implement the Android Code](#implement-the-android-code)
>    1. [Test the Native Module](#test-the-native-module)
> 1. [Create the Turbomodule](#create-the-turbomodule)
>    1. [Create the New Architecture Testing Environment](#create-the-new-architecture-testing-environment)
>    1. [Adding the JS Spec](#adding-the-js-spec)
>    1. [Generate the Code](#generate-the-code)
>    1. [Update the iOS Code](#update-the-ios-code)
>    1. [Update the Android Code](#update-the-android-code)
>    1. [Test the Turbomodule](#test-the-turbomodule)
> 1. [Summary](#summary)

Reviewer: ```suggestion
The guide is organized as it follows:

1. [What is a Turbomodule, In Short](#what-is-a-turbomodule-in-short)
1. [Create the Native Module](#create-the-native-module)
   1. [Creating the Structure](#creating-the-structure)
   1. [Create a Testing Environment](#create-a-testing-environment)
   1. [Define the JS Shared Code](#define-the-js-shared-code)
   1. [Implement the iOS Code](#implement-the-ios-code)
   1. [Implement the Android Code](#implement-the-android-code)
   1. [Test the Native Module](#test-the-native-module)
1. [Create the Turbomodule](#create-the-turbomodule)
   1. [Create the New Architecture Testing Environment](#create-the-new-architecture-testing-environment)
   1. [Adding the JS Spec](#adding-the-js-spec)
   1. [Generate the Code](#generate-the-code)
   1. [Update the iOS Code](#update-the-ios-code)
   1. [Update the Android Code](#update-the-android-code)
   1. [Test the Turbomodule](#test-the-turbomodule)
1. [Summary](#summary)
```

Can we remove the in-content ToC? We used those in the past, but since migration to Docusaurus V2, the ToC is generated automatically, so we stick to that. Also maintain manually the links in there, in a long term usually leads to the broken links anyway after few iteration of changes.

<img width="1098" alt="Screenshot 2022-03-21 123443" src="https://user-images.githubusercontent.com/719641/159253872-24a3d7c9-cf30-4820-b03a-c894cfcd29de.png">

<!-- To learn more about Turbomodule, check this [deep dive](). -->

## Create the Native Module

This section and the following are a practical guides to create a module that is backward compatible. We are not diving deep into what can be done with Native Modules. For more informations on that topic, have a look [here](native-modules-intro).

If you are interested in creating a Turbomodule, you can skip this part and directly go [here](#create-the-turbomodule). Consider that most of the code of a Native Module is needed for a Turbomodule.

### Creating the Structure

We have two alternatives to create the basic structure for our module.

1. Using the React Native CLI
2. Manually.

#### React Native CLI

To create a turbomodule with the React Native CLI, run the following command:

```sh
npx create-react-native-library <name-of-the-library>
```

> Follow the questions the CLI asks and, once the process is done, it will create a folder structure completed with an example app, typescript configuration and much more. The sample structure is shown in the following image:
> 
> <center><img src="docs/assets/Turbomodules/NpxLibFolderStructure.png" /></center>

Reviewer: ```suggestion
Follow the questions the CLI asks and, once the process is done, it will create a folder structure completed with an example app, typescript configuration and much more.

<figure>
  <img src="docs/assets/Turbomodules/NpxLibFolderStructure.png" />
  <figcaption>The sample directory structure.</figcaption>
</figure>
```

If the image needs a description, we can use `figure` and `figcaption` tags, which also is more semantic correct.

> This is a good starting point, but remember to double check what will be published if you keep this setup. Moreover, the Typescript support for the new architecture is still in beta, so you won't be able to use it for the Turbomodule sections.

Reviewer: ```suggestion
This is a good starting point, but remember to double check what will be published if you keep this setup. Moreover, the TypeScript support for the new architecture is still in beta, so you won't be able to use it for the Turbomodule sections.
```

```js
// @flow
// 1
import { NativeModules, Platform } from 'react-native';

// 2
const LINKING_ERROR =
  'The package "AuthorizationManager" doesn\'t seem to be linked. Make sure: \n\n' +
  Platform.select({
    ios: "- You have run 'pod install'\n",
    default: ''
  }) +
  '- You rebuilt the app after installing the package\n' +
  '- You are not using Expo managed workflow\n';

// 3
const manager = NativeModules.AuthorizationManager
  ? NativeModules.AuthorizationManager
  : new Proxy(
      {},
      {
        get() {
          throw new Error(LINKING_ERROR);
        }
      }
    );

// 4
export type AuthorizationStatus =
  | 'NotDetermined'
  | 'Restricted'
  | 'Denied'
  | 'Authorized'
  | 'AuthorizedAlways'
  | 'AuthorizedWhenInUse';

export default manager;
```

:::info
This file uses [`Flow`](https://flow.org/) to work with types.
Types are required for the Turbomodules
:::

The file imports the `NativeModules` and the `Platform` objects from `react-native` (marked with `// 1` in the code).
Then, it prepares the manager if the module is properly registered (`// 3`). Otherwise, when we try to use it, we get an error (set up in `// 2`).

> Finally, it exposes a type for the possible alternative authorization status and the manager object (`// 4`).

Reviewer: Can we put those comment directly in the code example? It would be more aligned with the other code examples in the docs.

:::important
The `ExampleLibrary.xcodeproj` must be in the `ios` folder. If you put it into a nested folder, module autodiscovery won't be able to find it.
:::

:::info
Coding in Xcode without installing the React dependencies could be tricky.
Our suggestion is to use the `OldArchitectureApp` created [here](#create-a-testing-environment) as test bed to have a working iOS environment. Then, remember to copy back to this folder all the code you write in the `OldArchitectureApp` native files.
Once the `ExampleLibrary.xcodeproj` has been created, have a look at the [Test the Native Module](#test-the-native-module) section to see how to install the pod.
:::

Open the `ELExampleLibrary.h` and replace its code with the following:

```objective-c
#import <React/RCTBridgeModule.h>

@interface ELExampleLibrary : NSObject <RCTBridgeModule>

@end
```

This code imports the `RCTBridgeModule.h` from the React Native library and it declares that the `ELExampleLibrary` will conform to the `RCTBridgeModule` protocol.

Then, open the `ELExampleLibrary.m` and replace the code with the following:

> ```obj-c

Reviewer: ```suggestion
```obj-c title="ELExampleLibrary.m"
```

We can use title to add filename to the codeblock, not required change but always nice to see. For more information check:
* https://docusaurus.io/docs/markdown-features/code-blocks#code-title

```

Before running the app, remember to add the required `Info.plist` keys:

- `NSLocationWhenInUseUsageDescription`
- `NSLocationAlwaysAndWhenInUseUsageDescription`

Now, from the terminal, run the `npx react-native run-ios` command to see the app working.

## Summary

To summarise, the steps to create a Turbomodule which is bacward compatible with the old architecture, are:

1. Create a Native Module.
2. Create the Turbomodule specs.
3. Make sure that the interfaces of the Native Module and of the Turbomodule are the same.
4. Update the module podspec, using the `ENV["RCT_NEW_ARCH_ENABLED"] == "1"` guard to avoid installing the dependencies in the old architecture
5. Add the `codegenConfig` to the `package.json` of the module
6. Add the `#import "<moduleName>Spec.h`, guarded by an `#ifdef RCT_NEW_ARCH_ENABLED` guard
7. Implement the `getTurboModule` C++ method in the native code, guarded by an `#ifdef RCT_NEW_ARCH_ENABLED` guard.

Then, in your app, you can choose which version of the module to use with the proper `import` in JS and installing the pods with the right command:

- `pod install` for the old architecture

> - `RCT_NEW_ARCH_ENABLED=1 pod install` for the new one

Reviewer: Probably we want add note for M1 users in there. There is a partial for that named `_markdown-m1-cocoapods.mdx`.
